coalib/core/PersistentHash.py

```diff
@@ -1,19 +1,42 @@
+from collections import Iterable
 from hashlib import sha1
 import pickle
 
 
-def persistent_hash(obj):
+def order(obj):
+    """
+    Recursively preserves the order of any sets or dicts
+    found in obj.
     """
-    Calculates a persistent hash of the given object.
+    if isinstance(obj, set):
+        return sorted(obj)
+    if isinstance(obj, dict):
+        temp_obj = obj
```

Comment:
hehe nope, that doesn't create a copy ;) all python objects are references, so you need to do a first-level copy (aka shallow copy) in this case (because you already work recursively here), see `copy.copy`. However, you need to be careful with dicts referencing each other, that would cause an infinite loop (please add a test for that case!).
Maybe you can also work with deep-copies, behaviour for deep-copying is overridable as well. So you can only make it copy dictionaries itself, but not any other objects, then you totally avoid recursion problems.

Reply:
Or maybe order the dict beforehand into a list/tuple structure, and then iterate over it and replace dict/set values with their ordered counterparts.